src/functional/iterator/transforms/popup_tmps.py

```diff
@@ -1,60 +1,141 @@
+        The returned `wrap` function undos the extraction in all cases; for example,
+        `wrap(f, args...)` returns `lift(f)(args...)` in the first case.
+        """
+        if isinstance(node.fun, ir.FunCall) and node.fun.fun == ir.SymRef(id="lift"):
+            # lifted lambda call or lifted scan
+            assert len(node.fun.args) == 1
+            fun = node.fun.args[0]
+
+            is_scan = isinstance(fun, ir.FunCall) and fun.fun == ir.SymRef(id="scan")
+            if is_scan:
+                assert isinstance(fun, ir.FunCall)  # just for mypy
+                fun = fun.args[0]
+
+            def wrap(fun: ir.Lambda, args: list[ir.Expr]) -> ir.FunCall:
+                if is_scan:
+                    assert isinstance(node.fun, ir.FunCall) and isinstance(
+                        node.fun.args[0], ir.FunCall
+                    )
+                    scan_args = [cast(ir.Expr, fun)] + node.fun.args[0].args[1:]
+                    f: Union[ir.Lambda, ir.FunCall] = ir.FunCall(
+                        fun=ir.SymRef(id="scan"), args=scan_args
+                    )
+                else:
+                    f = fun
+                return ir.FunCall(fun=ir.FunCall(fun=ir.SymRef(id="lift"), args=[f]), args=args)
+
+            assert isinstance(fun, ir.Lambda)
+            return fun, True, wrap
+        if isinstance(node.fun, ir.Lambda):
             # direct lambda call
-            lifts = dict()
-            res = self.generic_visit(node, lifts=lifts)
-            symbol_map = {param.id: arg for param, arg in zip(res.fun.params, res.args)}
-            new_args = [
-                RemapSymbolRefs().visit(arg, symbol_map=symbol_map) for arg in lifts.values()
-            ]
-            assert len(res.fun.params) == len(res.args + new_args)
-            return ir.FunCall(fun=res.fun, args=res.args + new_args)
 
-        return self.generic_visit(node, lifts=lifts)
+            def wrap(fun: ir.Lambda, args: list[ir.Expr]) -> ir.FunCall:
+                return ir.FunCall(fun=fun, args=args)
+
+            return node.fun, False, wrap
+
+        return None
 
-    def visit_Lambda(self, node: ir.Lambda, *, lifts):
-        node = self.generic_visit(node, lifts=lifts)
-        if not lifts:
-            return node
+    def visit_FunCall(
+        self, node: ir.FunCall, *, lifts: Optional[dict[ir.Expr, ir.SymRef]] = None
+    ) -> Union[ir.SymRef, ir.FunCall]:
+        if call_info := self._extract_lambda(node):
+            fun, is_lift, wrap = call_info
 
-        new_params = [ir.Sym(id=param) for param in lifts.keys()]
-        return ir.Lambda(params=node.params + new_params, expr=node.expr)
+            nested_lifts = dict[ir.Expr, ir.SymRef]()
+            fun = self.visit(fun, lifts=nested_lifts)
+            # Note: lifts in arguments are just passed to the parent node
+            args = self.visit(node.args, lifts=lifts)
+
+            if is_lift:
+                assert lifts is not None
```

Comment:
please add a unit test for this code path

Reply:
Done.